### ISSUE TITLE
practical testing of what will become borg 1.2

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#4360]
do practical testing with master branch code (or some alpha release).

report anything you find on the issue tracker (as a separate issue), check first if the issue already has been reported. you can post "works for me" in *this* ticket here.

see end of this ticket for current betas / rc (or directly check the github releases page).

**do not run master branch code or borg pre-releases against production repos, always use fresh repos made by `borg init`!**

**Comment by fantasya-pbem** [comment]
I get a strange error message when trying to backup / (fs root):
$ borg create /srv/testing::complete_manual_4 /
/: [Errno 2] No such file or directory: ''

But trying to backup /root for example works as expected. Is this a bug?

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem yes, that's a bug. can you please open a separate issue for it?

**Comment by fantasya-pbem** [comment]
Done: #4405

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/milestone/61 planned todo and known bugs for alpha 3.

**Comment by fantasya-pbem** [comment]
I have tested master branch (borg 1.2.0a3.dev3+g81f9a8cc) with empty repo on davfs2.
- borg init -e keyfile-blake2
- borg create --exclude-from $exclude --exclude-caches --compression lz4
- borg check
Everything worked perfectly.
```$ borg info
                       Original size      Compressed size    Deduplicated size
All archives:               41.16 GB             32.78 GB             31.31 GB
                       Unique chunks         Total chunks
Chunk index:                  194392               251139```

WORKS FOR ME

**Comment by ThomasWaldmann** [comment]
alpha 3 is out. \o/

**Comment by fantasya-pbem** [comment]
Tested latest master branch (borg 1.2.0a3.dev8+gde151cd3) with my 40 GB repo from yesterday on davfs2.

- borg create --exclude-from $exclude --exclude-caches --compression lz4 (~ 4 min)
- borg check (> 37 min)

No errors occurred. 87 Segments total.

```
                       Original size      Compressed size    Deduplicated size
All archives:               81.83 GB             65.08 GB             31.85 GB
                       Unique chunks         Total chunks
Chunk index:                  195405               501636
```
Interesting: davfs didn't respect cache_size of 5 GB, all segments were cached until cache was 30 GB. When borg exited (rc 0), cache was cleaned to < 5 GB, which is expected behaviour when open files are closed.

works for me

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem the age threshold of the new lrucache is 4min, so it would be interesting to test >>4min. Also, the code is avoiding to use old file handles, but otoh does **not** actively search for them and close them.

**Comment by fantasya-pbem** [comment]
I did several more tests in the last two days, with a production repository I used for half a year until some days ago:
- dedup size: 117 GB, 2233 segments (original size 7.6 TB)
- initial backup consists of 181 segments, ~77 GB dedup size

First some tests with 5 GB limited DavFS cache (which is on its own LVM volume that had ~45 GB free space).

As expected, borg 1.2.0a3.dev8+gde151cd3 crashed when DavFS cache was full (45 GB).
After that I tested a patched borg 1.1.10.dev5+g7a7e04a4.d20190227 from the 1.1-maint branch with deactivated LRU fd cache. borg check was fine and DavFS cache was constantly at 5 GB until segment 131 when it crashed with KeyError / Input/Output error.

Therefore I increased DavFS cache to 25 GB limit.

borg 1.1.10.dev5+g7a7e04a4.d20190227 check then finished without errors.

Now I tested borg 1.2.0a3.dev8+gde151cd3 again. DavFS deleted the "metadata files" from its cache but no segment files when the 25 GB threshold was exceeded.
**BUT** – when cache reached 40 GB (that was when segment file 100 was downloaded) DavFS started to delete segments (seg 1 first, then 2, ...). This was 31 minutes after I started the check.
Cache stayed at 40–41 GB until segment 182 (first seg of 2nd backup in repo). DavFS now deleted a bunch of files fast and could manage to hold 25 GB until end of borg check.

This is a very interesting behaviour, but I cannot see a plausible explaination why DavFS cache growth stopped at 41 GB after 31 minutes. I can only draw the conclusion that one needs a reasonably large DavFS cache for borg checks, but I does not need to be as big as the borg repo.

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem i opened #4427 for further work on the lrucache. if you like to help with testing, add a comment there.

**Comment by ThomasWaldmann** [comment]
Just released 1.2.0 alpha4 to pypi.

**Comment by fantasya-pbem** [comment]
Just for protocol: Tested borg-1.2.0a5 today. Works for me.

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem did you also test some of the new features (see changelog)?

**Comment by fantasya-pbem** [comment]
I did a "borg compact --cleanup-commits" which did what it should. I ran a "borg check" afterwards which succeeded too. (Test repo has 5 archives. Will do more tests later with a production repo copy.)

I did search for information about the borg check "--max-duration". It seems that this option is not explained in detail anywhere. I'm wondering what would be good and bad max duration values. There should be a section in the docs where this option is discussed.

**Comment by ThomasWaldmann** [comment]
`--max-duration` should be listed with a brief description in the `borg create` reference docs.

Good values depend a bit on your repo size and also on speeds and overheads. But I guess usually one would take enough time to make a decent amount of progress related to the overall time needed to check the whole repo.

E.g. if whole repo would take 100h to check, you could:
- run a 48h check at the weekend, when backup are only made made Mo..Fr
- run a 4h check each night, when every day backups are created and enough time for them is left.
- ...

**Comment by fantasya-pbem** [comment]
I tested "borg compact", "borg prune" and "borg check" with by big repo on DavFS. Compact and prune worked well (prune's applied-rule output is nice!). Check did crash after 90 segments, but I think this was because of small DavFS cache (5 G) which may be too less for the network throughput of my server and Borg's FD timeout. With 10 G cache borg check did succeed. So again – 1.2.0a5 works for me.

Regarding --max-duration: It is mentioned in "borg check" with one sentence: _"do only a partial repo check for max. SECONDS seconds (Default: unlimited)"_.
There is no further explaination why and when this option could be used. I assume that instead of one full long-running check Borg can run multiple partial checks that last --max-duration until the whole repo is checked. But this fact doesn't seem to be expressed with this clarity in the docs. It should be mentioned in borg check.

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem thanks for testing, created #4473 for the docs issue.

**Comment by ThomasWaldmann** [comment]
1.2.0a6 was just released - please test it: https://github.com/borgbackup/borg/releases/tag/1.2.0a6

This time the first time including pyinstaller-made binaries for Linux, FreeBSD and macOS, please test them (see the README about the binaries).

**Comment by christophlehmann** [comment]
I'm very happy to see the new `compact` command.

Server-Version 1.1.9
Client-Version: borg-linux64-1.2.0a6

I ran `borg compact $BORG_REPO` on client. it ended ~1 sec later and exit code was 0. The compact command is not available in 1.1.9 so i expect an error message and rc>0

**Comment by christophlehmann** [comment]
I ran `borg-linux64-1.2.0a6 compact` on the server. The server had no repokey. It shrinked the repository from 287 GB to 25 which seems reasonable. Then i ran a backup with `borg-linux64-1.1.10` which went fine. An extract with `borg-linux64-1.1.10` on the oldest backup also went fine.

**Comment by ThomasWaldmann** [comment]
@christophlehmann `borg compact` is executed on the server (even if you start it on the client).

Servers < 1.2 do not know `compact`, so yes, guess that should error.

Thanks for testing stuff!

**Comment by ThomasWaldmann** [comment]
> I ran borg compact $BORG_REPO on client. it ended ~1 sec later and exit code was 0.
> The compact command is not available in 1.1.9 so i expect an error message and rc>0

I checked this and (slightly counter-intuitive though), the "OK" rc (rc == 0) is correct.

This is because borg < 1.2 always did the compaction when committing the repo (so basically for most repo-writing / committing commands), so it did what you told it to do although < 1.2 does not support the separate compact command (which is rather about **not** doing the compaction in all other commands, but only doing it in the compact command).

**Comment by ntova** [comment]
Just finished testing my most common commands with version 1.2.0a6: `check`, `compact`, `create`, `delete`, `diff`, `extract`, `list`, `prune`, `recreate` against a copy of my main 110G repository which is by now about 1 1/2 years old and is used daily. Everything works as expected, no errors, and no huge changes performance-wise except:

The one thing I noticed, is that computing the stats now seems to take a bit longer, at least when using `recreate`.

As an example: 800MB repository with 158 archives on a Raspberry Pi 4, recreate over all archives to exclude a folder:

recreate times | borg 1.1.10 | borg 1.2.0a6
---- | ---- | ----
with stats | 6:45 min | 12:00 min + 4sec compact
w/o stats | 6:50 min | 3:47 min + 4sec compact

I know that computing the stats during recreate is unnecessary, I sometimes just like to see the numbers shrink ^^ The missing compact after each archive really shows, I like the separate compact behavior :)

**Comment by ThomasWaldmann** [comment]
That recreate --stats performance regression looks like an interesting thing to profile and find the root cause.

"borg info" and "--stats" had some bugs in 1.1, so some things are done differently in 1.2 to give correct results now, but having twice the runtime seems a bit much.

**Comment by ntova** [comment]
I just checked, the relevant commit is 61b9283567b18863ef66fd624c61299da0f130fd which makes sense. All archives in my test were pre 1.2, therefore the 1.2 metadata is not available and `--stats` is slow. A second run of recreate gives better performance.

WARNING: same repo, but new machine, so times are not comparable to my previous comment.

version | time for recreate (min:sec)
--- | ---
1.1.10 | 3:28
1.2.0a6, old metadata | 4:16
1.2.0a6, new metadata / second run | 2:40
1.2 before 61b9283567b18863ef66fd624c61299da0f130fd | 2:15

**Comment by szpak** [comment]
@ThomasWaldmann Do you plan to release a7 soon or it's better to build Borg from master to give it a try?

**Comment by ThomasWaldmann** [comment]
Yes, I should release a new alpha (and first look at / update the change log to see how much changed sinced last alpha). Hopefully in next few weeks.

**Comment by szpak** [comment]
I've seen updated changes, thanks. Are they some automatically generated binary snapshot versions (for Linux) generated from CI (I wasn't able to find any)? Or I need to setup the Python developer environment do build one (a wait "next few weeks")?

**Comment by ThomasWaldmann** [comment]
I don't trust cloud services enough to build distribution binaries there.

I am still waiting for the cooler for the 2nd cpu for my machine to arrive (thanks to DELL to use a non-standard cpu cooler mount), then I'll run the next vagrant tests / builds there.

**Comment by szpak** [comment]
I was thinking just about snapshots for non-critical stuff (testing). However, potentially malicious binaries could compromise the whole system, so I can find reasoning for it.
No need to hurry. In the meantime I have found [instructions](https://github.com/borgbackup/borg/blob/master/docs/development.rst#using-vagrant) how to use vagrant, do probably I will be able to generate a statically linked Linux binary for my own purpose.

**Comment by ThomasWaldmann** [comment]
Good news: coolers arrived, dual-xeon workstation now happily powering borg testing with 16C/32T. \o/

So, next alpha "coming soon".

**Comment by ThomasWaldmann** [comment]
1.2.0a7 released just now! please test. https://github.com/borgbackup/borg/releases/tag/1.2.0a7

**Comment by bket** [comment]
Just started testing 1.2.0a7 on OpenBSD. So far no surprises when:
- Initialising new repositories with, and without encryption;
- Creating some snapshots to these repositories;
- Restoring some files;
- Listing snapshots;
- Compacting.

I really like that the first ctrl-c makes a checkpoint and then aborts.

Also, regression tests run successfully.

**Comment by bket** [comment]
I tested 1.2.0a7 with msgpack-0.6.2, which seems to work ok.

After reading #3757 I understand that I need to be carful. Are there specific tests that need to be conducted before a pull request is even considered?

**Comment by infectormp** [comment]
@bket here some info about msgpack compatibilities https://github.com/borgbackup/borg/issues/4220

**Comment by ThomasWaldmann** [comment]
The 0.6.1 -> 0.6.2 changes are not many / big and also do not look problematic:

https://github.com/msgpack/msgpack-python/commits/master

So guess we can just accept 0.6.2 as compatible until proven else.

Guess the bundled msgpack should also get upgraded to 0.6.2 (only in master branch, 1.1-maint branch is still at 0.5.x) as it contains necessary changes for bundling now in the upstream src.

**Comment by henfri** [comment]
Hello,

I would like to test the windows version. Would you be able to provide a compiled windows version as well?
If not, instructions in the Readme_windows would be appreciated on how to use the non compiled version (no need to start at installing python ;-) 

Regards,
Hendrik

**Comment by henfri** [comment]
Hm neither?

**Comment by ThomasWaldmann** [comment]
@henfri ask @jrast about the current windows work.

**Comment by henfri** [comment]
Hello Thomas,
I thought I was doing that.
Do you mean in a different issue?

Greetings
Hendrik

**Comment by ThomasWaldmann** [comment]
Mentioning him might be enough to notify him.

**Comment by jrast** [comment]
Hey @henfri 
I'm currently working (or was working last month) on CI for Windows, so there are some builds at appveyor which should work: https://ci.appveyor.com/project/jrast/borg/build/artifacts

Not that a wheel and a exe is provided. The wheel is expected to be installable with pip without any further dependencies, the exe should be a single file executable which does not require a python installation on your machine.

The CI work is done in this PR: https://github.com/borgbackup/borg/pull/4733

**Comment by dutchgecko** [comment]
Just leaving a comment to note that running a 1.2.0a7 client under WSL seems to work without having to invoke the `workarounds` env variable. I've only run a create and check so far, are there any other behaviours that are worth giving a poke?

#### Client info

* borg version: `borg 1.2.0a7`
* `cat /etc/issue`: `Ubuntu 19.04`
* `uname -a`: `Linux mortred 4.4.0-18362-Microsoft #1-Microsoft Mon Mar 18 12:02:00 PST 2019 x86_64 x86_64 x86_64 GNU/Linux`
* Windows version: 1903; build: 18362.418

#### Server info

* borg version: `borg 1.2.0a7`
* `cat /etc/issue`: `Welcome to Alpine Linux 3.10`
* `uname -a`: `Linux 1526ba0a7d8f 5.2.0-3-amd64 #1 SMP Debian 5.2.17-1 (2019-09-26) x86_64 Linux`
* docker version: `Docker version 19.03.4, build 9013bf583a`

#### Used commands

* `borg create`
* `borg prune` [note: there was nothing to prune]
* `borg check`

**Comment by ThomasWaldmann** [comment]
@dutchgecko nobody needs `create`, everybody needs `extract` (just joking).

Also: create some archives, delete or prune some, run a full repo check after some operations.

**Comment by theChaosCoder** [comment]
@jrast I just tried your borg.exe. Simply calling borg.exe throws this error:

```
Traceback (most recent call last):
  File "borg\__main__.py", line 14, in <module>
  File "c:\projects\borg\borg-env\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
  File "borg\__init__.py", line 5, in <module>
ModuleNotFoundError: No module named 'borg._version'
[14168] Failed to execute script __main__
```

**Comment by ThomasWaldmann** [comment]
@theChaosCoder @jrast that is a file usually generated by setuptools_scm (see `setup_requires` in `setup.py`). It cares for creating a version from git tags and also for the manifest.

**Comment by jrast** [comment]
@theChaosCoder can you provide the link to the exact *.exe which you have downloaded and used? This seems to be related to the build process.

**Comment by theChaosCoder** [comment]
I clicked on https://ci.appveyor.com/project/jrast/borg/build/artifacts and there's only one borg.exe
Direct link: 
https://ci.appveyor.com/api/buildjobs/x169rd0r1db5xh24/artifacts/dist%2Fborg.exe

**Comment by jrast** [comment]
@theChaosCoder can you try again? I rebased the my repo on the current master which triggered a new build. On my machine the exe did startup as expected.

**Comment by theChaosCoder** [comment]
Yep, this one starts with no errors.

**Comment by szpak** [comment]
TL;TR. I've been playing with 1.2.0a7 for a while and I haven't encounter any regression.

I have been backing up ~10TB of various data (several archives to deal with different retention) locally (to an another HDD), side by side (two separate repos) with 1.1.10 for a few weeks. Occasionally I was comparing content (bytes and basic metadata) of the mounted archive with the original source filed (or the other backup). There were no issues with data corruption. The archives were slightly smaller in 1.2 (auto,zstd), but for 2 of them the output was bigger (however, it was stable).

The biggest improvement was of course in backuping up to devices with slower random access time (pendrive, sshfs) thanks to relaxed need to do compacting after every archive (tested with smaller data set). With multiple archives I achieved 4-6 times shorter backup time. It's a killer features for those environments!

**Comment by m3nu** [comment]
Also added the latest 1.2.0a7 as option to [BorgBase.com](https://www.borgbase.com) for testing. Just choose that version in the repo options under advanced.

I'll update it whenever I see a new release on Github.

![Screen Shot 2020-04-15 at 15 52 40](https://user-images.githubusercontent.com/3916435/79314042-d0b7fd80-7f33-11ea-96c4-e0688a2921a4.jpg)

**Comment by theChaosCoder** [comment]
On Windows10 x64 I tried to make a test backup with the latest win build
https://ci.appveyor.com/project/jrast/borg/builds/32112933/artifacts
```
.\borg.exe create  D:\repotest::abc D:\myfolder\                                                                      Traceback (most recent call last):
  File "borg\__main__.py", line 15, in <module>
  File "borg\archiver.py", line 4492, in main
  File "borg\archiver.py", line 4297, in get_args
  File "borg\archiver.py", line 4343, in parse_args
  File "argparse.py", line 1755, in parse_args
  File "argparse.py", line 1787, in parse_known_args
  File "argparse.py", line 1996, in _parse_known_args
  File "argparse.py", line 1952, in consume_positionals
  File "argparse.py", line 1861, in take_action
  File "argparse.py", line 1158, in __call__
  File "argparse.py", line 1787, in parse_known_args
  File "argparse.py", line 1996, in _parse_known_args
  File "argparse.py", line 1952, in consume_positionals
  File "argparse.py", line 1845, in take_action
  File "argparse.py", line 2375, in _get_values
  File "argparse.py", line 2408, in _get_value
  File "borg\helpers\parseformat.py", line 503, in validator
  File "borg\helpers\parseformat.py", line 390, in __init__
  File "borg\helpers\parseformat.py", line 396, in parse
  File "borg\helpers\parseformat.py", line 421, in _parse
NameError: name 'path' is not defined
[60140] Failed to execute script __main__
```

**Comment by ThomasWaldmann** [comment]
@theChaosCoder Maybe you want to test #4862 by @jrast ?

**Comment by ThomasWaldmann** [comment]
@m3nu cool!

I am trying to make a8 soon, but currently so much is broken in master branch's vagrant machines that it is blocking that.

**Comment by m3nu** [comment]
> I am trying to make a8 soon, but currently so much is broken in master branch's vagrant machines that it is blocking that.

You mean for building the Pyinstaller binaries? Happy to have a look. Which ones are having issues exactly?

**Comment by ThomasWaldmann** [comment]
There are some "vagrant"-labelled tickets. Some of them need a specific box though.

**Comment by jrast** [comment]
@theChaosCoder Sorry, somehow one line got lost in the windows path handling PR... Anyway, I submited a new PR which fixes this: https://github.com/borgbackup/borg/pull/5119
Can you check with the binaries provided there?

**Comment by theChaosCoder** [comment]
Is must say I didn't expect the win version to work this well. I tried create, list, info, delete. Everything works. Even smb shares 
`.\borg create --stats -p -v --list --filter=AMEx  -C auto,lzma,7 --exclude-from   \\qnap\Backups\borg_exclude.txt  d:\gam::Games_win  \\qnap\Games`

**So far I found only 2 small problems:** (Everything tested on Win10 x64 powershell)
the exclude file looks like this: 
`\\qnap\Games\@Recycle`
`\\qnap\*\@Recycle`
But borg still adds the files from the Recycle folder. It's a smb share folder from my qnap nas.
My cygwin version looks like this and works correctly:
`//qnap/*/@Recycle`

The "info msg" does not stay at the bottom like in a linux terminal / cygwin 
`987.68 MB O 875.50 MB C 360 B D 112 N  ....`
Idk know if it would be an easy task or not to display it on windows like in a linux terminal

@jrast Have you tried to connect via ssh with win borg?
`.\borg create  --remote-path /opt/bin/borg  ssh:\\admin@qnap:22\share\Backups\BackupBorg::wintest  D:\test`

I get this error. But I don't know how the correct format would look like.
`Invalid location format: "ssh:\\admin@qnap:22\share\Backups\BackupBorg::wintest"`

**Comment by jrast** [comment]
@theChaosCoder Thanks for testing! Yea, basic backup / list / info / delete / restore should work! 

**Some Important Notes:**
- This is a early, early alpha for windows! Do not use it, under any circumstances, on productive data!
- currently absolute path are stored in the repo, so archives created on windows can not be restored under Linux!
- Filtering seems to be broken. There are many tests in the testsuite which fail.
- Yep, currently only local paths are supported on windows. I will look into this in the future.
- There are many cases which are not tested or are not working! But now as basic support for windows paths is in the master branch I can continue to work on the next steps!

**Comment by theChaosCoder** [comment]
Borg.exe check also works, but I can't abort it with CTRL+C (forgot verbose param)

**Comment by fantasya-pbem** [comment]
@theChaosCoder 
> `Invalid location format: "ssh:\\admin@qnap:22\share\Backups\BackupBorg::wintest"`

Shouldn't the URL use / instead?

ssh://admin@qnap:22...

**Comment by theChaosCoder** [comment]
I tried \ and /
I need to figure out what "kind" of ssh on my system is running. It should be the windows10 native one, but I'm not 100% sure yet.

ssh output
usage: ssh [-46AaCfGgKkMNnqsTtVvXxYy] [-B bind_interface]
           [-b bind_address] [-c cipher_spec] [-D [bind_address:]port]
           [-E log_file] [-e escape_char] [-F configfile] [-I pkcs11]
           [-i identity_file] [-J [user@]host[:port]] [-L address]
           [-l login_name] [-m mac_spec] [-O ctl_cmd] [-o option] [-p port]
           [-Q query_option] [-R address] [-S ctl_path] [-W host:port]
           [-w local_tun[:remote_tun]] destination [command]

**Comment by jrast** [comment]
@theChaosCoder @fantasya-pbem it won't work either way! Only Windows Paths are accepted on Windows! See: https://github.com/borgbackup/borg/blob/master/src/borg/helpers/parseformat.py#L425

**Comment by jrast** [comment]
@theChaosCoder you can check the SSH version with `ssh -V`.

**Comment by jrast** [comment]
Getting SSH to work on windows is not as easy as expected. The RemoteRepository makes heavy use of `select.select` on stdin / stdout / stderr which makes totaly sense on Unix Systems. But silly Windows does not support select on files but only on sockets...

**Comment by theChaosCoder** [comment]
`ssh -V` => `OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5`

**Comment by m3nu** [comment]
Thanks for the update, @ThomasWaldmann !

Here my report from macOS:

1. When starting there is an error due to missing notarization.
![Screen Shot 2020-04-23 at 10 24 57](https://user-images.githubusercontent.com/3916435/80052483-5992f280-854d-11ea-8418-a01d30f4bbed.jpg)

2. To get around it the user can allow an exception in *Preferences > Security & Privacy*
![Screen Shot 2020-04-23 at 10 26 32](https://user-images.githubusercontent.com/3916435/80052540-7deecf00-854d-11ea-8f93-ebea99a7794a.jpg)

3. With the exception added, the first startup takes about 20 seconds due to validation by `syspolicyd` (to scan libraries online). Subsequent runs take a more manageable 3-4 seconds.
![Screen Shot 2020-04-23 at 10 27 34](https://user-images.githubusercontent.com/3916435/80052582-96f78000-854d-11ea-9658-599e0d9caebd.jpg)

**Comment by ThomasWaldmann** [comment]
@m3nu are macOS users familiar with doing these extra steps to allow stuff from unknown developers? if so, I guess we would not need to document that.

**Comment by m3nu** [comment]
Currently probably not, but advanced users will get familiar over time. So let's not document it for now.

**Comment by fantasya-pbem** [comment]
Currently testing 1.2.0a8 with small repo size but checking all command. Found a minor bug (#5159).

**Comment by dragetd** [comment]
> @m3nu are macOS users familiar with doing these extra steps to allow stuff from unknown developers? if so, I guess we would not need to document that.

This Mac behavior was introduced late 2019 (apps without signature no longer simply executable). Various other applications had the same issue, thus some users know about this and there are guides online. Yet, I would still recommend mentioning it in the docs.

Signing apps requires an Apple developer account which costs 99$ per year.

**Comment by m3nu** [comment]
> Signing apps requires an Apple developer account which costs 99$ per year.

I have this account for Vorta and other projects. But pyinstaller single binary apps can't be signed. I already tried that. Since Pyinstaller extracts all the libs to a temporary folder. So the signing doesn't carry over.

Probably not much to do for now. If Thomas agrees, I could add 1-2 sentences to the docs anyways.

**Comment by FabioPedretti** [comment]
> for linux, there are 2 binaries: one made on debian stretch, one on buster.

Out of curiosity, is there a special meaning for providing two linux 64 binaries, this time?
Thanks

**Comment by ThomasWaldmann** [comment]
So that both builds can be tested.
I hope they both work, but I am not sure *where* (on which OSes of what age) they work.
Depending on when 1.2 will be finally released, it might be either one of these builds.

**Comment by FabioPedretti** [comment]
Thank you for the feedback. I quickly tried both on Debian buster and they seems to work fine.

**Comment by fantasya-pbem** [comment]
On Ubuntu 18.04 and Linux Mint 19.3 `borg-buster64` doesn't work ("GLIBC_2.28 not found"), but borg-linux64 works as usual.

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/releases/tag/1.2.0a9 please help testing.

**Comment by m3nu** [comment]
Latest alpha also available on [BorgBase.com](https://www.borgbase.com) for testing as of now. Select `1.2.x` under *Advanced > BorgBackup Version*.
<img src="https://user-images.githubusercontent.com/3916435/95706203-cbd8ca00-0c88-11eb-909a-bef058b1316f.jpg" width="400" />

**Comment by noplanman** [comment]
ARM binaries are now built for alpha and beta versions as well.

https://borg.bauerj.eu/?show-unstable#archive

**Comment by ThomasWaldmann** [comment]
I just released 1.2.0b1 - please help testing and give feedback here.

**Comment by noplanman** [comment]
@ThomasWaldmann ARM builds are already available at https://borg.bauerj.eu/?show-unstable

**Comment by dietmargoldbeck** [comment]
Works fine for me on OpenBSD 6.8 amd64. I installed  1.2.0b1 with pip3.8 and run a backup of my OpenBSD test machine 
and a small test restore.
Thanks a lot for borg.

**Comment by m3nu** [comment]
Also available on [BorgBase.com](https://www.borgbase.com) as "testing" version.

<img src="https://user-images.githubusercontent.com/3916435/103117860-52f1e780-46a7-11eb-9077-27825dbebf77.jpg" width="250" />

**Comment by dietmargoldbeck** [comment]
Tested a  backup of my photo archive and filled up the filesystem. Borg stopped with filesystem full error.
Client is 1.2.0b1 in ubuntu and server is 1.2.0b1 on debian. Both using https://github.com/borgbackup/borg/releases/download/1.2.0b1/borg-linux64

Now there is a 470GB of borg but borg list is **empty**. 

I never filled up a filesystem with my production use of borg 1.1 and can't compare.
How can i test a restore from this archive? I actually wanted to test to restore most of my pictures, although the hard drive has a few badblocks triggering a firmware bug not to remap them.

Unfortunately borg check does not tell me which file read failed:
`dietmar@haubenpinguin:~$ borg-1.2 check --debug -v -p zwergpinguin-test:borg12test
Remote: Starting repository check
Traceback (most recent call last):

  File "borg/remote.py", line 239, in serve

  File "borg/repository.py", line 1008, in check

  File "borg/repository.py", line 1470, in iter_objects

  File "borg/repository.py", line 1560, in _read

OSError: [Errno 5] Input/output error

Borg server: Platform: Linux zwergpinguin-test 4.19.0-13-amd64 #1 SMP Debian 4.19.160-2 (2020-11-28) x86_64
Borg server: Linux: debian 10.7 
Borg server: Borg: 1.2.0b1  Python: CPython 3.7.9 msgpack: 1.0.0 fuse: llfuse 1.3.8 [pyfuse3,llfuse]
Borg server: PID: 23169  CWD: /home/dietmar
Borg server: sys.argv: ['borg', 'serve', '--info']
Borg server: SSH_ORIGINAL_COMMAND: None
Platform: Linux haubenpinguin 5.4.0-58-generic #64-Ubuntu SMP Wed Dec 9 08:16:25 UTC 2020 x86_64
Linux: debian bullseye/sid 
Borg: 1.2.0b1  Python: CPython 3.7.9 msgpack: 1.0.0 fuse: llfuse 1.3.8 [pyfuse3,llfuse]
PID: 302735  CWD: /home/dietmar
sys.argv: ['borg-1.2', 'check', '--debug', '-v', '-p', 'zwergpinguin-test:borg12test']
SSH_ORIGINAL_COMMAND: None
`

**Comment by infectormp** [comment]
@dietmargoldbeck borg say **OSError: [Errno 5] Input/output error** 
https://borgbackup.readthedocs.io/en/stable/faq.html#i-get-an-integrityerror-or-similar-what-now

**Comment by ThomasWaldmann** [comment]
@dietmargoldbeck looks like there are 2 issues with your repo:
- as the backup did not finish, you likely only have a checkpoint archive, try `borg list --consider-checkpoints repo`.
- you have some lower-level issue (fs, blockdevice, hdd/ssd) triggering that IOError when it does the repository check. that is below borg, so try to find out what is causing that.

**Comment by lfam** [comment]
I tested the update in GNU Guix. Everything seems fine for us.

**Comment by ThomasWaldmann** [comment]
I just merged #5574 - it would be interesting if that would get some testing in different scenarios / with different buffer sizes.

**Comment by anhofmann** [comment]
I have a repo that works fine with borg 1.1.11, but trying to do a `borg info` with 1.2.0b1 produces a traceback:
```
Local Exception
Traceback (most recent call last):
  File "borg/archiver.py", line 4736, in main
  File "borg/archiver.py", line 4667, in run
  File "borg/archiver.py", line 179, in wrapper
  File "borg/archiver.py", line 1406, in do_info
  File "borg/archiver.py", line 1488, in _info_repository
  File "borg/cache.py", line 407, in __str__
  File "borg/cache.py", line 434, in format_tuple
  File "borg/cache.py", line 426, in stats
  File "borg/archive.py", line 607, in calc_stats
  File "borg/archive.py", line 602, in add
  File "src/borg/hashindex.pyx", line 299, in borg.hashindex.ChunkIndex.__getitem__
KeyError: b'\xf4\xa2\xd4/\xacTs\xe0\xb6\xa6S\xe8B\x959\xb3\xac\xe8~\n\n\x9fu\xb2&Xp\n2\xda\xd84'

Platform: Linux rclone-archiv-ersteller 5.4.0-62-generic #70-Ubuntu SMP Tue Jan 12 12:45:47 UTC 2021 x86_64
Linux: debian bullseye/sid 
Borg: 1.2.0b1  Python: CPython 3.7.9 msgpack: 1.0.0 fuse: llfuse 1.3.8 [pyfuse3,llfuse]
```

I don't know what additional info I should provide, please ask, I will supply it.

**Comment by ThomasWaldmann** [comment]
@anhofmann can you give a command sequence which i could use to reproduce (from scratch, without having that repo)?

**Comment by dietmargoldbeck** [comment]
With the github Linux binary on my notebook and freebsd on the borg-server i get a Traceback at the end of the backup

Creating archive at "dgo-beastie:/backup/borgtest/archiv::brillenpinguin.full.20210129_1611902626"
/var/log/journal/585d4d9f8f3c47f888e7beb8be86f70e/system.journal: file changed while we backed it up
Local Exception
Traceback (most recent call last):
  File "borg/archiver.py", line 4736, in main
  File "borg/archiver.py", line 4667, in run
  File "borg/archiver.py", line 181, in wrapper
  File "borg/archiver.py", line 657, in do_create
  File "borg/archiver.py", line 628, in create_inner
  File "borg/cache.py", line 407, in __str__
  File "borg/cache.py", line 434, in format_tuple
  File "borg/cache.py", line 426, in stats
  File "borg/archive.py", line 607, in calc_stats
  File "borg/archive.py", line 602, in add
  File "src/borg/hashindex.pyx", line 299, in borg.hashindex.ChunkIndex.__getitem__
KeyError: b'\x93\xbb2\x8a\xa2-\xa9\xf1Qe\xe3N\xb8\xe6\x7f\xbfw\xcd\xf6\xf7\xebKys\xfajG\xbf|\xb4\xd8\x1d'

Platform: Linux brillenpinguin 5.10.9-201.fc33.x86_64 #1 SMP Wed Jan 20 16:56:23 UTC 2021 x86_64
Linux: Fedora 33 Thirty Three
Borg: 1.2.0b1  Python: CPython 3.7.9 msgpack: 1.0.0 fuse: llfuse 1.3.8 [pyfuse3,llfuse]
PID: 85929  CWD: /root/borgbackup/script
sys.argv: ['borg-1.2', 'create', '--stats', '--show-rc', '-C', 'lz4', '-c', '900', '--lock-wait', '7200', '-v', 'dgo-beastie:/backup/borgtest/archiv::brillenpinguin.full.20210129_1611902626', '/', '--exclude', '/proc', '--exclude', '/sys', '--exclude', '/dev', '--exclude', '/run', '--exclude', '/tmp', '--exclude', '/var/tmp', '--exclude', '/var/lib/libvirt', '--exclude', '/var/www/html/ftp', '--exclude', '/backup', '--exclude', '/media', '--exclude', '/zfs_usbdisk2-88411', '--exclude', '/backup-z1e721b5']
SSH_ORIGINAL_COMMAND: None

**Comment by ThomasWaldmann** [comment]
@dietmargoldbeck the easiest explanations could be:
- the repo was made and used by borg < 1.1.11
- the repo was made and used by an older 1.2 borg with the same bug
- you have faulty RAM - can you run memtest86+?

**Comment by dietmargoldbeck** [comment]
> @dietmargoldbeck the easiest explanations could be:
> 
>     * the repo was made and used by borg < 1.1.11

new repo just for beta testing. 
 
>     * the repo was made and used by an older 1.2 borg with the same bug

I only used the latest beta binaries Borg: 1.2.0b1  Python: CPython 3.7.9 msgpack: 1.0.0 fuse: llfuse 1.3.8 [pyfuse3,llfuse]

>     * you have faulty RAM - can you run memtest86+?

All machines survive mprime and are running borg-1.1 without problems. 
 
I could reproduce it with a checkpoint archive and continuing from my other notebook. Since it is easy to reproduce i opend
https://github.com/borgbackup/borg/issues/5668
I did not check yet if borg 1.1.15 is also affected.

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/releases/tag/1.2.0b2 just made a new release, please help testing!

**Comment by FabioPedretti** [comment]
With 1.2.0b2 borg info shows an error* and get stuck until I kill it. Using a repo created with previous 1.1 release.
*Note that I am using BORG_FILES_CACHE_SUFFIX, so the "files" file doesn't exist, but there are many "files.archive-number" files.

With borg 1.1.15:
```
# borg info
Repository ID: xxx
Location: /mnt/yyy/borg-repo
Encrypted: Yes (repokey BLAKE2b)
Cache: /mnt/yyy/borg-cache/xxx
Security dir: /root/.config/borg/security/xxx
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
All archives:              570.08 TB            433.07 TB              2.45 TB

                       Unique chunks         Total chunks
Chunk index:                 5142003           1124036218
```
With borg 1.2.0b2:
```
# ./borg-1.2.0b2 info
The files cache can't be read. [[Errno 2] No such file or directory: '/mnt/yyy/borg-cache/xxx/files']
Continuing without files cache - expect lower performance.
Repository ID: xxx
Location: /mnt/yyy/borg-repo
Encrypted: Yes (repokey BLAKE2b)
Cache: /mnt/yyy/borg-cache/xxx
Security dir: /root/.config/borg/security/xxx
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
^C^C^C^CKeyboard interrupt
```

Also it would be nice to explain in the changelog the difference between borg-linux64 and borg-buster64

**Comment by ThomasWaldmann** [comment]
@FabioPedretti hah, that was quick! :-)

The binaries are described in the README.txt right besides where you got the binaries. Basically it is stretch vs. buster.
stretch is still the default as it also supports older linux dists (older glibc), but testing buster would be also good because that will be next after stretch runs out of support.

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/pull/5433/files that was the PR for FILES_CACHE_SUFFIX and master branch. Have to check whether some place was forgotten to get updated with the `files_cache_name()` call.

**Comment by FabioPedretti** [comment]
Indeed no, I just set the env var during the create of every specific archive, but not during info, because I suppose the files cache just get used during creation.
Also borg info should give info about all the repo, so why I should set the env var?
Indeed with 1.1 it works fine, I use the env var only during create, and borg info shows no error.

**Comment by ThomasWaldmann** [comment]
@FabioPedretti you could open a ticket with all infos so we can investigate why 1.2 info opens the files cache (and so we can remove the above posts from here).

**Comment by FabioPedretti** [comment]
Done, see #5673.

**Comment by m3nu** [comment]
I noticed that `--log-json` isn't respected any more for archive progress. Reason being that it doesn't get passed to `Statistics()` via `FilesystemObjectProcessors()`. Already did a PR #5675 to fix it. (This also fixes my comment from Nov: https://github.com/borgbase/vorta/issues/238#issuecomment-722944599)

Before: (progress statistics are printed in last line)
![Screen Shot 2021-02-07 at 09 56 39](https://user-images.githubusercontent.com/3916435/107134346-47dbd980-692c-11eb-954b-537ab8913b68.jpg)

After: (progress is printed as json of type `archive_progress` again)
![Screen Shot 2021-02-07 at 09 58 15](https://user-images.githubusercontent.com/3916435/107134360-5e823080-692c-11eb-8dce-cade9a1e29b9.jpg)

**Comment by dietmargoldbeck** [comment]
With the latest beta i get tracebacks. It seems to happen when the ssh connection breaks due to keep alive problems (my internet connection is a bit unstable currently) and i could reproduce with rebooting the IPsec gateway which caused ssh keep alive failure.

`Creating archive at "dgo-beastie:/backup/borgtest/archiv::haubenpinguin.full.20210211_1613066629"
Remote: Timeout, server 10.129.48.9 not responding.ar/Pic...2190607/dsc07324.arw
Local Exception
Traceback (most recent call last):
  File "borg/archiver.py", line 181, in wrapper
  File "borg/archiver.py", line 661, in do_create
  File "borg/archiver.py", line 602, in create_inner
  File "borg/archiver.py", line 805, in _rec_walk
  File "borg/archiver.py", line 805, in _rec_walk
  File "borg/archiver.py", line 805, in _rec_walk
  [Previous line repeated 4 more times]
  File "borg/archiver.py", line 772, in _rec_walk
  File "borg/archiver.py", line 674, in _process_any
  File "borg/archive.py", line 1342, in process_file
  File "borg/archive.py", line 1177, in process_file_chunks
  File "borg/archive.py", line 1165, in chunk_processor
  File "borg/cache.py", line 920, in add_chunk
  File "borg/remote.py", line 476, in do_rpc
  File "borg/remote.py", line 709, in call
  File "borg/remote.py", line 826, in call_many
borg.remote.ConnectionClosed: Connection closed by remote host

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "borg/archiver.py", line 4798, in main
  File "borg/archiver.py", line 4729, in run
  File "borg/archiver.py", line 181, in wrapper
  File "borg/remote.py", line 634, in __exit__
  File "borg/remote.py", line 476, in do_rpc
  File "borg/remote.py", line 709, in call
  File "borg/remote.py", line 779, in call_many
  File "borg/remote.py", line 719, in send_buffer
  File "borg/remote.py", line 417, in write
BrokenPipeError: [Errno 32] Broken pipe

Platform: Linux haubenpinguin 5.4.0-65-generic #73-Ubuntu SMP Mon Jan 18 17:25:17 UTC 2021 x86_64
Linux: Unknown Linux  
Borg: 1.2.0b2  Python: CPython 3.9.1 msgpack: 1.0.2 fuse: llfuse 1.4.1 [pyfuse3,llfuse]
PID: 190358  CWD: /root/borgbackup/log
sys.argv: ['borg-1.2', 'create', '--stats', '--show-rc', '-C', 'lz4', '-c', '900', '--lock-wait', '7200', '-v', '-p', 'dgo-beastie:/backup/borgtest/archiv::haubenpinguin.full.20210211_1613066629', '/etc', '/root', '/usr/local', '/home', '--exclude', '/proc', '--exclude', '/sys', '--exclude', '/dev', '--exclude', '/tmp', '--exclude', '/var/tmp', '--exclude', '/run', '--exclude', '/snap', '--exclude', '/var/www/html/ftp', '--exclude', '/media', '--exclude', '/backup', '--exclude', '/mnt']
SSH_ORIGINAL_COMMAND: None

`
Should i make a bug entry? Any further info needed?

**Comment by ThomasWaldmann** [comment]
@dietmargoldbeck if the connection breaks down, there is not much borg can do. we maybe could catch the exception and make a more "pretty" error message, but when doing that, there is always some danger of hiding problems or making problems more difficult to debug, thus I rather tend not to do that.

**Comment by ThomasWaldmann** [comment]
May 2021 I released https://github.com/borgbackup/borg/releases/tag/1.2.0b3 - did anybody test it practically, any issues?

**Comment by m3nu** [comment]
[BorgBase.com](https://www.BorgBase.com) users can choose to use a beta version for their repo. 8 repos use it and 5 of those get daily backups. Didn't get any complaints there.

**Comment by bket** [comment]
Tested 1.2.0b3 on OpenBSD amd64 current:
- Builds OK (using an updated port)
- Regression tests OK
- Run tested (init/create/prune/compact/list/extract) OK

**Comment by Rudd-O** [comment]
Happy to see progress towards borg 1.2.

**Comment by ThomasWaldmann** [comment]
BTW, I plan to do another beta (b4) within the next days, please test it rather quickly after it is released. Too many code changes since b3 to directly go to rc1.

There will be only a relatively short time for testing b4 because I plan to release rc1 and 1.2.0 final (== rc1 + bug fixes, if any) in february.

**Comment by ThomasWaldmann** [comment]
1.2.0b4 is out and i also created a related discussion there:
https://github.com/borgbackup/borg/discussions/6166

please help testing, i want to do rc1 in about 2 weeks.

**Comment by dragetd** [comment]
When installing via pip (tested on debian 11, had a report on fedora):

```
root@9758c6dacda9:/# pip install borgbackup==1.2.0b4
Collecting borgbackup==1.2.0b4
  Using cached borgbackup-1.2.0b4.tar.gz (4.0 MB)
  Preparing metadata (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-b3uo71sm/borgbackup_47b698927b9e4a2e89ed2808a8f0d41b/setup.py'"'"'; __file__='"'"'/tmp/pip-install-b3uo71sm/borgbackup_47b698927b9e4a2e89ed2808a8f0d41b/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-llrftfmj
       cwd: /tmp/pip-install-b3uo71sm/borgbackup_47b698927b9e4a2e89ed2808a8f0d41b/
  Complete output (8 lines):
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/pip-install-b3uo71sm/borgbackup_47b698927b9e4a2e89ed2808a8f0d41b/setup.py", line 185, in <module>
      setup_crypto.crypto_ext_kwargs(pc, system_prefix_openssl),
    File "/tmp/pip-install-b3uo71sm/borgbackup_47b698927b9e4a2e89ed2808a8f0d41b/setup_crypto.py", line 32, in crypto_ext_kwargs
      raise Exception('Could not find OpenSSL lib/headers, please set BORG_OPENSSL_PREFIX')
  Exception: Could not find OpenSSL lib/headers, please set BORG_OPENSSL_PREFIX
  Warning: can not import pkgconfig python package.
  ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/c8/9b/23c53cee4ead57c8f92b134dc537d58583599b29558bd8c3b6ba26bb00fd/borgbackup-1.2.0b4.tar.gz#sha256=f39a4e013a61146fdaebbf71ec9fa727aa6f24d3bf5ae1330ecf4f7151b48fc0 (from https://pypi.org/simple/borgbackup/) (requires-python:>=3.8). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement borgbackup==1.2.0b4 (from versions: 0.23.0, 0.24.0, 0.25.0, 0.26.0, 0.26.1, 0.27.0, 0.28.0, 0.28.1, 0.28.2, 0.29.0, 0.30.0, 0.30.1, 1.0.0rc1, 1.0.0rc2, 1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.0.4, 1.0.5, 1.0.6rc1, 1.0.6, 1.0.7rc1, 1.0.7rc2, 1.0.7, 1.0.8rc1, 1.0.8, 1.0.9rc1, 1.0.9, 1.0.10rc1, 1.0.10, 1.0.11rc1, 1.0.11, 1.0.12, 1.0.13, 1.1.0b1, 1.1.0b2, 1.1.0b3, 1.1.0b4, 1.1.0b5, 1.1.0b6, 1.1.0rc1, 1.1.0rc2, 1.1.0rc3, 1.1.0rc4, 1.1.0, 1.1.1, 1.1.2, 1.1.3, 1.1.4, 1.1.5, 1.1.6, 1.1.7, 1.1.8, 1.1.9, 1.1.10, 1.1.11, 1.1.12, 1.1.13, 1.1.14, 1.1.15, 1.1.16, 1.1.17, 1.2.0a2, 1.2.0a3, 1.2.0a4, 1.2.0a5, 1.2.0a6, 1.2.0a7, 1.2.0a8, 1.2.0a9, 1.2.0b1, 1.2.0b2, 1.2.0b3, 1.2.0b4)
ERROR: No matching distribution found for borgbackup==1.2.0b4
```

Installing `python3-openssl` or `libssl-dev` did not fix it. Single executable binary works.

**Comment by ThomasWaldmann** [comment]
@dragetd can you try again, but first `pip install pkgconfig`?

**Comment by dragetd** [comment]
I am testing using a debian container, launched via:
`docker run -it --rm debian`
and installing borg using the command
`apt update && apt install -y pip pkg-config libssl-dev && pip install -U pip setuptools wheel && pip install pkgconfig && pip install borgbackup==1.2.0b4`

Still has issues installing. The result is now: http://rn0.me/show/M5qD2g25GlI5Qbk9N2F8/

I continued to figure out the missing package: `pgk-config' `libssl-dev` and `libacl1-dev` and it finally worked.

Can we document these dependencies a bit better on the installation page maybe?

**Comment by ThomasWaldmann** [comment]
we have a whole section in the docs that talks about all the libs.
the only thing that needs checking is whether it also talks about the `pip install pkgconfig` (which is new in 1.2).

**Comment by ThomasWaldmann** [comment]
note: i updated the docs with "pip install pkgconfig" (and also some other stuff) recently.

**Comment by noplanman** [comment]
Just a quick note that the most recent ARM builds are ready for testing too, up until 1.2.0b4:

https://borg.bauerj.eu/?show-unstable

**Comment by ThomasWaldmann** [comment]
1.2.0rc1 is released - please test!

1.2.0 final will have the same bugs as rc1 except if you find and report them in the next 2 weeks. :-)

**Comment by fantasya-pbem** [comment]
I maintain a 1.1.17 repository containing daily backups of some server's root file system over a couple of years now, with a pruning policy that keeps 12 month. This repo has > 70 GB. I made a 1:1 copy of it and use that for a parallel daily backup, testing 1.2.0rc1. I also duplicated the backup client with a different hostname, so `.cache/borg` and `.config.borg` are the same.

The client mounts its repository via SSHFS, then runs Borg commands on the mounted repo. This is done because I have SSH user access to the backup server, but cannot install the Borg binary there.

The first `borg list` on this repo copy was made with 1.1.17 and I got the expected "was previously located" message, which I accepted. I then started testing with a simple `borg list` and `borg info`.

First thing I noted: `borg list` works like before, but a `borg info` on the latest backup now takes around 70 seconds, where 1.1.17 only took around 2-3 seconds. It seems that Borg 1.2 reads a lot of data over the SSH connection. Is this expected behaviour in my case?

**Comment by bket** [comment]
Tested 1.2.0rc1 on OpenBSD amd64 current:

- Builds OK (using an updated port)
- Regression tests OK
- Tested local and remote repositories, and all kind of operations, e.g. create/prune/compact/list/extract, wihout  issues

**Comment by FabioPedretti** [comment]
> I maintain a 1.1.17 repository containing daily backups of some server's root file system over a couple of years now, with a pruning policy that keeps 12 month. This repo has > 70 GB. I made a 1:1 copy of it and use that for a parallel daily backup, testing 1.2.0rc1. I also duplicated the backup client with a different hostname, so `.cache/borg` and `.config/borg` are the same.
> 
> The client mounts its repository via SSHFS, then runs Borg commands on the mounted repo. This is done because I have SSH user access to the backup server, but cannot install the Borg binary there.
> 
> The first `borg list` on this repo copy was made with 1.1.17 and I got the expected "was previously located" message, which I accepted. I then started testing with a simple `borg list` and `borg info`. I have not executed the `borg compact --cleanup-commits` command yet.
> 
> First thing I noted: `borg list` works like before, but a `borg info` on the latest backup now takes around 70 seconds, where 1.1.17 only took around 2-3 seconds. It seems that Borg 1.2 reads a lot of data over the SSH connection. Is this expected behaviour in my case?

I have a similar issue, also with a repo on SSHFS, see #5804.

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem @FabioPedretti if you make a 1:1 copy, the repository ID will be identical. borg uses that to access some clientside configs/caches, so **your 1.1 borg and your 1.2 borg now will use different repos, but same clientside config/cache storage, not good**.

I did the same thing btw, but before ever accessing the copy, I changed the repo id (edit repo_dir/config).

What you should do now is:
- stop all borg processes
- edit the repo id in the borg1.2 repo copy
- borg1.1 delete --cache-only REPO_ORIG
- borg1.2 delete --cache-only REPO_COPY
- that will cause a cache resync when the borg run next time, but at least make sure the cache is ok (is separate) again. not sure about stuff in `.config` (e.g. nonce).

**Comment by ThomasWaldmann** [comment]
(and yes, i need to look at borg info...)

**Comment by fantasya-pbem** [comment]
> your 1.1 borg and your 1.2 borg now will use different repos, but same clientside config/cache storage, not good

I am aware of that. In my case, Borg 1.1 and 1.2 run on different clients (clones). They do not share the same cache, so the identical ID should not be an issue, as they do not interfere with each other.

**Comment by ThomasWaldmann** [comment]
Upgrade notes: https://github.com/borgbackup/borg/pull/6257/files

**Comment by m3nu** [comment]
Version `1.2.0rc1` is now available for **new** repos on [BorgBase.com](https://www.borgbase.com) in all regions when selecting "Borg 1.2.x (Testing)" as version from *Advanced* options.

Noticed that pruning is much faster and shows the corresponding rule. That's very helpful. 👍

Regarding the slow `borg create --stats` command, there won't be a change? (mentioned here https://github.com/borgbackup/borg/discussions/6166#discussioncomment-2040866). As you explained, the `--stats` command is very slow for repos created with Borg 1.1.x. The client or server version used doesn't seem to matter. Is there any way to "upgrade" a repo? If there are good reasons to keep it as-is, we'll adjust Vorta to work without `--stats` option. Already prepared a PR [here](https://github.com/borgbase/vorta/pull/1196).

**Comment by m3nu** [comment]
> borg compact does not need a key, so it is possible to invoke it from the client or also from the server.

This is also pretty exciting, since we can run it during off-peak hours and clean up people's repos. Already tested and no password was needed. Will this also remove data left after pruning append-only repos when run server-side in full access mode? I didn't see anything explicit in the command doc or notes.

**Comment by fantasya-pbem** [comment]
Executed `borg compact --cleanup-commits`, and `borg check` was fine without errors found. Creating a new backup also worked well. Only `borg info` is still very slow.

**Comment by ThomasWaldmann** [comment]
@m3nu borg compact behaves the same no matter from where you invoke it (on client or on server).

It will just physically remove everything that logically is already deleted. So delete and prune basically mark stuff as deleted, but the data is still physically present in old segment files (but these old segment files have "logical holes" where this deleted data is stored, because there is already a later DELETE committed for that chunk of data).

borg compact then cleans up and removes all these logical holes be writing only the stuff that is still in use to new segment files (but skipping the deleted stuff). there is a default threshold so that this data shuffling does not occur if there is too little to compact.

So, delete and prune do not immediately free space. But a compact done afterwards will.

And compact will reject to run if the repo is in append-only mode.

**Comment by ThomasWaldmann** [comment]
@m3nu yeah, good idea, borg hosting providers could offer off-peak-hours compacting:
- let the user enable this (default disabled, because the user might want to have temporary-append-only behaviour)
- alternatively, let the user forbid it
- consider timezones, maybe run compact at weekend-everywhere (like between Sat 12:01 .. Sun 11:59 UTC)?
- clever scheduling reduces hdd head thrashing
- if user runs out of quota, run compact if not forbidden.

**Comment by ThomasWaldmann** [comment]
About the create --stats speed: let's continue in #6259.

**Comment by m3nu** [comment]
> whether the append-only is enforced by repo config or by `borg serve`

Should be fine, if ‘borg compact’ is rejected for both options. I see the main difference in *how* both options are enforced. Server-side, vs user-side. The latter is like making my own file RO to avoid deleting it. Stops accidents, but doesn’t add security.

Also thanks for the additional background. Will add compaction as opt-in feature.

**Comment by fantasya-pbem** [comment]
I noticed that 1.2 now prints out a warning like _„file changed while we backed it up“_, which seems to be the cause for return code 1 (warning). I never experienced that in 1.1. This might be generated because I add `--progress` or `--stats` but this changed behaviour might be considered a breaking change.

**Comment by ThomasWaldmann** [comment]
@fantasya-pbem well, a warning is not an error and warnings are kind of expected if you back up an active filesystem.

i don't remember whether we have the same code in 1.1 (or backported the change from master to 1.1-maint), but master/1.2 checks the file stat before and after reading an input file and notices if it has changed while we read it (which can mean that we backed up inconsistent crap).  iirc tar also emits a similar warning for this case.

for files you do not really care about inconsistency might be acceptable, but borg does not known whether you do or not. you can also exclude unimportant changing files if your goal is a warning free run.

**Comment by ThomasWaldmann** [comment]
OK, just wanted to tell that there is about 1 week left before 1.2.0 official release.

Also, all tickets in 1.2.0 milestone (except "testing it", this one) were closed or moved to future milestones.

So, please use the remaining time for more stress testing and feedback.

**Comment by fantasya-pbem** [comment]
In the last couple of days I had one backup per day, a `borg prune`, a `borg compact` and a check afterwards. Everything without issues.

**Comment by ThomasWaldmann** [comment]
had to do some late code changes in #6301 and #6306, hopefully no regressions.

**Comment by fantasya-pbem** [comment]
Maybe release one more RC, esp. for #6306? I cannot estimate the impact, is this fix relevant for special use cases? ("race condition")

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/tree/1.2.0 i just tagged 1.2.0 in the repo.

if somebody has a development setup for borg and wants to give the release some pre-release testing, that's the tag to fetch.

i still have to do the platform testing (using vagrant), so the tag might change to another changeset later if anything needing a fix comes up.

**Comment by ThomasWaldmann** [comment]
https://paste.thinkmo.de/7ikN34mK#borg-1.2.0-packages here are the currently planned release files for 2022-02-22.

Please give them a try before I officially upload them to github and pypi.

**Comment by bket** [comment]
Once again tested on OpenBSD:
- builds, and passes all regression tests
- tested against fresh, and a couple of 'production' repos. Tried several options all of which behave as expected

**Comment by ThomasWaldmann** [comment]
@LocutusOfBorg @FelixSchwarz ^^^

**Comment by FelixSchwarz** [comment]
@ThomasWaldmann I'll try to get Fedora's spec file updated and run a test build this evening. Might be just in time before beta cutoff for F36 (Feb 22).

**Comment by ThomasWaldmann** [comment]
Last minute! :)

**Comment by FelixSchwarz** [comment]
Tested on Fedora rawhide:

- builds ok, test suite passes
- creating a simple repo works (+ restore)

What I'd need for real Fedora build is a tarball (+ ideally a GPG signature). I generated a tarball manually from a git checkout due to setuptools_scm issues but that won't fly for a real distro package.

**Comment by szpak** [comment]
@FelixSchwarz It would be good to it confirmed by @ThomasWaldmann , but the aforementioned [link](https://paste.thinkmo.de/7ikN34mK#borg-1.2.0-packages) contains a (digitally signed) tarball - borgbackup-1.2.0.tar.gz (second from the bottom) which - as I understood will be the official one (assuming no disasters reported :-) ).

**Comment by FelixSchwarz** [comment]
@szpak Thanks, I completely missed the link.

**Comment by ThomasWaldmann** [comment]
Yeah, that is the plan: release that stuff "as is" if no major issues are reported.

**Comment by LocutusOfBorg** [comment]
Hello, I'll have a test too

**Comment by LocutusOfBorg** [comment]
I did test on Debian sid and Ubuntu jammy (openssl 3.0.1). Build was fine after adding some new dependencies (dateutil, pkgconfig).
I got some "deprecated warnings" with new openssl, but nothing that prevented build/tests from working.
```
src/borg/platform/posix.c: In function '__pyx_pf_4borg_8platform_5posix_2swidth':
src/borg/platform/posix.c:1502:3: warning: 'PyUnicode_AsUnicode' is deprecated [-Wdeprecated-declarations]
 1502 |   __pyx_t_2 = __Pyx_PyUnicode_AsUnicode(__pyx_v_s); if (unlikely((!__pyx_t_2) && PyErr_Occurred())) __PYX_ERR(0, 19, __pyx_L1_error)
      |   ^~~~~~~~~
In file included from /usr/include/python3.9/unicodeobject.h:1026,
                 from /usr/include/python3.9/Python.h:93,
                 from src/borg/platform/posix.c:19:
/usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
  580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
      |                                             ^~~~~~~~~~~~~~~~~~~
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -g -fwrapv -O2 -Wl,-z,relro -Wl,-z,now -g -O2 -ffile-prefix-map=/borgbackup=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 build/temp.linux-x86_64-3.9/src/borg/platform/posix.o -o /borgbackup/.pybuild/cpython3_3.9/build/borg/platform/posix.cpython-39-x86_64-linux-gnu.so
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -O2 -ffile-prefix-map=/borgbackup=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -Isrc/borg/crypto -I/usr/include/python3.9 -c src/borg/crypto/low_level.c -o build/temp.linux-x86_64-3.9/src/borg/crypto/low_level.o
src/borg/crypto/low_level.c: In function '__pyx_pf_4borg_6crypto_9low_level_10_AEAD_BASE_10decrypt':
src/borg/crypto/low_level.c:10318:149: warning: passing argument 4 of 'EVP_CIPHER_CTX_ctrl' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
10318 |     __pyx_t_6 = ((!(EVP_CIPHER_CTX_ctrl(__pyx_v_self->ctx, EVP_CTRL_GCM_SET_TAG, __pyx_v_self->mac_len, (((unsigned char const *)__pyx_v_idata.buf) + __pyx_v_hlen)) != 0)) != 0);
      |                                                                                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
In file included from src/borg/crypto/low_level.c:644:
/usr/include/openssl/evp.h:684:71: note: expected 'void *' but argument is of type 'const unsigned char *'
  684 | int EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr);
      |                                                                 ~~~~~~^~~
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -g -fwrapv -O2 -Wl,-z,relro -Wl,-z,now -g -O2 -ffile-prefix-map=/borgbackup=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 build/temp.linux-x86_64-3.9/src/borg/crypto/_crypto_helpers.o build/temp.linux-x86_64-3.9/src/borg/crypto/low_level.o -lcrypto -o build/lib.linux-x86_64-3.9/borg/crypto/low_level.cpython-39-x86_64-linux-gnu.so
building 'borg.compress' extension
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -O2 -ffile-prefix-map=/borgbackup=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -Isrc/borg -I/usr/include/python3.9 -c src/borg/compress.c -o build/temp.linux-x86_64-3.9/src/borg/compress.o
src/borg/compress.c: In function '__pyx_pf_4borg_8compress_4ZSTD_2_decide':
src/borg/compress.c:6113:31: warning: comparison of integer expressions of different signedness: 'size_t' {aka 'long unsigned int'} and 'int' [-Wsign-compare]
 6113 |   __pyx_t_2 = ((__pyx_v_osize < __pyx_v_isize) != 0);
      |                               ^
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -g -fwrapv -O2 -Wl,-z,relro -Wl,-z,now -g -O2 -ffile-prefix-map=/borgbackup=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 build/temp.linux-x86_64-3.9/src/borg/compress.o -llz4 -lzstd -o build/lib.linux-x86_64-3.9/borg/compress.cpython-39-x86_64-linux-gnu.so
building 'borg.hashindex' extension
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -O2 -ffile-prefix-map=/borgbackup=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -Isrc/borg -I/usr/include/python3.9 -c src/borg/hashindex.c -o build/temp.linux-x86_64-3.9/src/borg/hashindex.o
In file included from src/borg/cache_sync/cache_sync.c:19,
                 from src/borg/hashindex.c:642:
src/borg/cache_sync/unpack.h: In function 'unpack_callback_array_end':
src/borg/cache_sync/unpack.h:276:70: warning: pointer targets in passing argument 2 of 'hashindex_get' differ in signedness [-Wpointer-sign]
  276 |         cache_entry = (uint32_t*) hashindex_get(u->chunks, u->current.key);
      |                                                            ~~~~~~~~~~^~~~
      |                                                                      |
      |                                                                      char *
In file included from src/borg/hashindex.c:641:
src/borg/_hashindex.c:552:54: note: expected 'const unsigned char *' but argument is of type 'char *'
  552 | hashindex_get(HashIndex *index, const unsigned char *key)
      |                                 ~~~~~~~~~~~~~~~~~~~~~^~~
In file included from src/borg/cache_sync/cache_sync.c:19,
                 from src/borg/hashindex.c:642:
src/borg/cache_sync/unpack.h:290:52: warning: pointer targets in passing argument 2 of 'hashindex_set' differ in signedness [-Wpointer-sign]
  290 |             if(!hashindex_set(u->chunks, u->current.key, cache_values)) {
      |                                          ~~~~~~~~~~^~~~
      |                                                    |
      |                                                    char *
In file included from src/borg/hashindex.c:641:
src/borg/_hashindex.c:562:54: note: expected 'const unsigned char *' but argument is of type 'char *'
  562 | hashindex_set(HashIndex *index, const unsigned char *key, const unsigned char *value)
      |                                 ~~~~~~~~~~~~~~~~~~~~~^~~
In file included from src/borg/cache_sync/cache_sync.c:19,
                 from src/borg/hashindex.c:642:
src/borg/cache_sync/unpack.h:290:58: warning: passing argument 3 of 'hashindex_set' from incompatible pointer type [-Wincompatible-pointer-types]
  290 |             if(!hashindex_set(u->chunks, u->current.key, cache_values)) {
      |                                                          ^~~~~~~~~~~~
      |                                                          |
      |                                                          uint32_t * {aka unsigned int *}
In file included from src/borg/hashindex.c:641:
src/borg/_hashindex.c:562:80: note: expected 'const unsigned char *' but argument is of type 'uint32_t *' {aka 'unsigned int *'}
  562 | hashindex_set(HashIndex *index, const unsigned char *key, const unsigned char *value)
      |                                                           ~~~~~~~~~~~~~~~~~~~~~^~~~~
src/borg/hashindex.c: In function '__pyx_pf_4borg_9hashindex_7NSIndex_2__setitem__':
src/borg/hashindex.c:4680:95: warning: passing argument 3 of 'hashindex_set' from incompatible pointer type [-Wincompatible-pointer-types]
 4680 |   __pyx_t_5 = ((!(hashindex_set(__pyx_v_self->__pyx_base.index, ((unsigned char *)__pyx_t_4), __pyx_v_data) != 0)) != 0);
      |                                                                                               ^~~~~~~~~~~~
      |                                                                                               |
      |                                                                                               uint32_t * {aka unsigned int *}
In file included from src/borg/hashindex.c:641:
src/borg/_hashindex.c:562:80: note: expected 'const unsigned char *' but argument is of type 'uint32_t *' {aka 'unsigned int *'}
  562 | hashindex_set(HashIndex *index, const unsigned char *key, const unsigned char *value)
      |                                                           ~~~~~~~~~~~~~~~~~~~~~^~~~~
src/borg/hashindex.c: In function '__pyx_pf_4borg_9hashindex_10ChunkIndex_2__setitem__':
src/borg/hashindex.c:6027:95: warning: passing argument 3 of 'hashindex_set' from incompatible pointer type [-Wincompatible-pointer-types]
 6027 |   __pyx_t_5 = ((!(hashindex_set(__pyx_v_self->__pyx_base.index, ((unsigned char *)__pyx_t_4), __pyx_v_data) != 0)) != 0);
      |                                                                                               ^~~~~~~~~~~~
      |                                                                                               |
      |                                                                                               uint32_t * {aka unsigned int *}
In file included from src/borg/hashindex.c:641:
src/borg/_hashindex.c:562:80: note: expected 'const unsigned char *' but argument is of type 'uint32_t *' {aka 'unsigned int *'}
  562 | hashindex_set(HashIndex *index, const unsigned char *key, const unsigned char *value)
      |                                                           ~~~~~~~~~~~~~~~~~~~~~^~~~~
src/borg/hashindex.c: In function '__pyx_f_4borg_9hashindex_10ChunkIndex__add':
src/borg/hashindex.c:7818:80: warning: passing argument 3 of 'hashindex_set' from incompatible pointer type [-Wincompatible-pointer-types]
 7818 |     __pyx_t_1 = ((!(hashindex_set(__pyx_v_self->__pyx_base.index, __pyx_v_key, __pyx_v_data) != 0)) != 0);
      |                                                                                ^~~~~~~~~~~~
      |                                                                                |
      |                                                                                uint32_t * {aka unsigned int *}
In file included from src/borg/hashindex.c:641:
src/borg/_hashindex.c:562:80: note: expected 'const unsigned char *' but argument is of type 'uint32_t *' {aka 'unsigned int *'}
  562 | hashindex_set(HashIndex *index, const unsigned char *key, const unsigned char *value)
      |                                                           ~~~~~~~~~~~~~~~~~~~~~^~~~~
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -g -fwrapv -O2 -Wl,-z,relro -Wl,-z,now -g -O2 -ffile-prefix-map=/borgbackup=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 build/temp.linux-x86_64-3.9/src/borg/hashindex.o -o build/lib.linux-x86_64-3.9/borg/hashindex.cpython-39-x86_64-linux-gnu.so
src/borg/platform/posix.c: In function '__pyx_pf_4borg_8platform_5posix_2swidth':
src/borg/platform/posix.c:1502:3: warning: 'PyUnicode_AsUnicode' is deprecated [-Wdeprecated-declarations]
 1502 |   __pyx_t_2 = __Pyx_PyUnicode_AsUnicode(__pyx_v_s); if (unlikely((!__pyx_t_2) && PyErr_Occurred())) __PYX_ERR(0, 19, __pyx_L1_error)
      |   ^~~~~~~~~
In file included from /usr/include/python3.9/unicodeobject.h:1026,
                 from /usr/include/python3.9/Python.h:93,
                 from src/borg/platform/posix.c:19:
/usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
  580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
      |                                             ^~~~~~~~~~~~~~~~~~~
creating build/lib.linux-x86_64-3.9/borg/platform

```

**Comment by ThomasWaldmann** [comment]
IIRC, `dateutil` is only needed for the tests.

When "pyx" shows up in warnings: that is code generated by Cython and besides using the latest Cython release, we can't do much about it.

_hashindex.c is hand-made C code, we can look at whether we can get rid of some warnings (e.g. by adding casts), but as far as I have seen yet, warnings are harmless.

**Comment by ThomasWaldmann** [comment]
1.2.0 is released - let's continue in the related discussion / new github tickets.
